tests/query/processors/test_tuple_elementer.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+import pytest
+
+from snuba.query.dsl import identity as dsl_identity
+from snuba.query.dsl import literals_tuple, tupleElement
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    Expression,
+    FunctionCall,
+    Lambda,
+    Literal,
+)
+from snuba.query.processors.tuple_elementer import TupleElementer
+from snuba.request.request_settings import HTTPRequestSettings
+from snuba.state import set_config
+from tests.query.processors.query_builders import build_query
+
+
+def equals(op1: Expression, op2: Expression, alias: str | None = None) -> FunctionCall:
+    return FunctionCall(alias, "eq", tuple([op1, op2]))
+
+
+def some_tuple(*tuple_elems):
+    return literals_tuple(None, [Literal(None, e) for e in tuple_elems])
+
+
+def identity(expression: Expression) -> Expression:
+    return dsl_identity(expression, None)
+
+
+def arrayjoin_pattern():
+    """The TupleElementer does not support calculated tuples. One such place
+    these occur is the arrayjoin optimizer. They look like this:
+
+    tupleElement(
+        arrayJoin(
+          arrayMap(
+            (x,y ->
+              tuple(
+                x,
+                y
+              )
+            ),
+            tags.key,
+            tags.value
+          )
+        ) AS `snuba_all_tags`,
+        2
+      ) AS `_snuba_tags_value` |> tags_value
+    The TupleElementer should not touch them
+    """
+    return FunctionCall(
+        alias="_snuba_tags_key",
+        function_name="tupleElement",
+        parameters=(
+            FunctionCall(
+                alias="snuba_all_tags",
+                function_name="arrayJoin",
+                parameters=(
+                    FunctionCall(
+                        alias=None,
+                        function_name="arrayMap",
+                        parameters=(
+                            Lambda(
+                                alias=None,
+                                parameters=("x", "y"),
+                                transformation=FunctionCall(
+                                    alias=None,
+                                    function_name="tuple",
+                                    parameters=(
+                                        Argument(alias=None, name="x"),
+                                        Argument(alias=None, name="y"),
+                                    ),
+                                ),
+                            ),
+                            Column(alias=None, table_name=None, column_name="tags.key"),
+                            Column(
+                                alias=None, table_name=None, column_name="tags.value"
+                            ),
+                        ),
+                    ),
+                ),
+            ),
+            Literal(alias=None, value=1),
+        ),
+    )
+
+
+TEST_QUERIES = [
```

Comment:
Should we add test cases for curried_functions and lambda's as well since this visitor has implementations for them?